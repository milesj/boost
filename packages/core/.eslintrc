{
  "rules": {
    "@typescript-eslint/no-explicit-any": "off",
    "@typescript-eslint/no-unsafe-assignment": "off"
  }
}
