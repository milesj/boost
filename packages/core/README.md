# ðŸš€ Boost

[![Build Status](https://github.com/milesj/boost/workflows/Build/badge.svg)](https://github.com/milesj/boost/actions?query=branch%3Amaster)
[![npm version](https://badge.fury.io/js/%40boost%2Fcore.svg)](https://www.npmjs.com/package/@boost/core)
[![npm deps](https://david-dm.org/milesj/boost.svg?path=packages/core)](https://www.npmjs.com/package/@boost/core)

A cross-platform framework for building robust Node.js command line applications and developer
tools.

## Features

- Common utilities, classes, and patterns.
- Args parsing and formatting.
- Logging, debugging, and crash reporting.
- Event listening and emitting.
- Serial and parallel based pipelines.
- Built-in localization and translations.
- And more!

## Requirements

- Node 8.9+

## Documentation

[https://milesj.gitbook.io/boost](https://milesj.gitbook.io/boost)
