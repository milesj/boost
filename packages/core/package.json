{
  "name": "@boost/core",
  "version": "1.18.6",
  "description": "Robust pipeline for creating dev tools that separate logic into routines and tasks.",
  "keywords": [
    "boost",
    "build",
    "tool",
    "pipeline",
    "routine",
    "task",
    "config"
  ],
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "nonce": "1579986563745",
  "engines": {
    "node": ">=8.9.0"
  },
  "repository": "https://github.com/milesj/boost/tree/master/packages/core",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@boost/common": "^1.8.5",
    "@boost/debug": "^1.4.6",
    "@boost/event": "^1.3.1",
    "@boost/internal": "^1.2.0",
    "@boost/log": "^1.2.0",
    "@boost/terminal": "^1.0.3",
    "@boost/translate": "^1.3.9",
    "@types/debug": "^4.1.5",
    "@types/yargs": "^15.0.4",
    "@types/yargs-parser": "^15.0.0",
    "debug": "^4.1.1",
    "env-ci": "^4.5.1",
    "execa": "^3.4.0",
    "exit": "^0.1.2",
    "fast-glob": "^3.2.2",
    "fs-extra": "^8.1.0",
    "json5": "^2.1.3",
    "lodash": "^4.17.15",
    "optimal": "^4.2.0",
    "pluralize": "^8.0.0",
    "split": "^1.0.1",
    "yargs-parser": "^18.1.2"
  },
  "funding": {
    "type": "ko-fi",
    "url": "https://ko-fi.com/milesjohnson"
  }
}
