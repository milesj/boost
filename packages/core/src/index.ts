/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import CLI from './CLI';
import ConfigLoader from './ConfigLoader';
import Console from './Console';
import Context from './Context';
import Emitter, { EventArguments, EventListener } from './Emitter';
import ModuleLoader from './ModuleLoader';
import Pipeline from './Pipeline';
import Plugin from './Plugin';
import Reporter from './Reporter';
import Routine from './Routine';
import Task from './Task';
import Tool, { ToolOptions, ToolConfig, ToolPluginRegistry } from './Tool';

export {
  CLI,
  ConfigLoader,
  Console,
  Context,
  Emitter,
  EventArguments,
  EventListener,
  ModuleLoader,
  Pipeline,
  Plugin,
  Reporter,
  Routine,
  Task,
  Tool,
  ToolConfig,
  ToolOptions,
  ToolPluginRegistry,
};

export * from './types';
