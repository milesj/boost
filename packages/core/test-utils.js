module.exports = require('./lib/testUtils');
