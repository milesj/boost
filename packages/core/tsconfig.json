{
  "compilerOptions": {
    "declarationDir": "lib",
    "outDir": "lib",
    "rootDir": "src"
  },
  "exclude": [
    "lib",
    "tests"
  ],
  "extends": "../../tsconfig.options.json",
  "include": [
    "src/**/*",
    "types/**/*",
    "../../types/**/*"
  ],
  "references": [
    {
      "path": "../common"
    },
    {
      "path": "../debug"
    },
    {
      "path": "../event"
    },
    {
      "path": "../internal"
    },
    {
      "path": "../log"
    },
    {
      "path": "../translate"
    }
  ]
}