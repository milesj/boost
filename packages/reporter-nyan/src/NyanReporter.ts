/* eslint-disable no-magic-numbers, @typescript-eslint/no-explicit-any */

import { style } from '@boost/terminal';
import { Reporter, Routine, Task } from '@boost/core';

// Based on the wonderful reporter found in Mocha
// https://github.com/mochajs/mocha/blob/master/lib/reporters/nyan.js
export default class NyanReporter extends Reporter {
  activeRoutine: Routine<any, any> | null = null;

  activeTask: Task<any> | null = null;

  catWidth: number = 11;

  colorIndex: number = 0;

  failed: boolean = false;

  rainbows: string[][] = [[], [], [], []];

  rainbowColors: number[] = [];

  rainbowWidth: number = 0;

  tick: 1 | 2 | 3 | 4 = 1;

  blueprint() {
    return {};
  }

  bootstrap() {
    super.bootstrap();

    this.rainbowWidth = this.size().columns - this.catWidth * 2;
    this.rainbowColors = this.generateColors();

    this.console.onStart.listen(this.handleStart);
    this.console.onStop.listen(this.handleStop);
    this.console.onRoutine.listen(this.handleRoutine);
    this.console.onTask.listen(this.handleTask);
  }

  handleStart = () => {
    this.createOutput(() => this.renderLines()).enqueue();
    this.hideCursor();
  };

  handleStop = (error: Error | null) => {
    this.failed = !!error;
    this.showCursor();
  };

  handleRoutine = (routine: Routine<any, any>) => {
    const handler = () => {
      this.activeRoutine = routine;
    };

    routine.onFail.listen(handler);
    routine.onPass.listen(handler);
    handler();
  };

  handleTask = (task: Task<any>) => {
    this.activeTask = task;
  };

  applyColor(text: string): string {
    if (!this.hasColorSupport()) {
      return text;
    }

    const color = this.rainbowColors[this.colorIndex % this.rainbowColors.length];

    // eslint-disable-next-line unicorn/escape-case
    return `\u001b[38;5;${color}m${text}\u001b[0m`;
  }

  generateColors(): number[] {
    const colors: number[] = [];

    for (let i = 0; i < 6 * 7; i += 1) {
      const pi3 = Math.floor(Math.PI / 3);
      const n = i * (1 / 6);
      const r = Math.floor(3 * Math.sin(n) + 3);
      const g = Math.floor(3 * Math.sin(n + 2 * pi3) + 3);
      const b = Math.floor(3 * Math.sin(n + 4 * pi3) + 3);

      colors.push(36 * r + 6 * g + b + 16);
    }

    return colors;
  }

  getCatFace(): string {
    const routine = this.activeRoutine;
    let eyes = '- .-';

    if (routine) {
      if (routine.hasFailed()) {
        eyes = 'x .x';
      } else if (routine.isSkipped() || routine.isPending()) {
        eyes = 'o .o';
      } else if (routine.hasPassed()) {
        eyes = '^ .^';
      }
    }

    if (this.failed) {
      eyes = 'x .x';
    }

    return `( ${style.white(eyes)})`;
  }

  increaseRainbowWidth() {
    for (let i = 0; i < 4; i += 1) {
      const line = this.rainbows[i];

      if (line.length >= this.rainbowWidth) {
        line.shift();
      }

      line.push(this.applyColor(this.isInterval() ? '_' : '-'));

      this.rainbows[i] = line;
    }

    this.colorIndex += 1;
  }

  increaseTick() {
    this.tick += 1;

    if (this.tick === 5) {
      this.tick = 1;
    }
  }

  isInterval(): boolean {
    return this.tick >= 3;
  }

  renderLines(): string {
    this.increaseTick();
    this.increaseRainbowWidth();

    const tick = this.isInterval();
    let output = '\n';

    // Poptart
    output += this.rainbows[0].join('');
    output += ' ';
    output += style.magenta(' ,------,');
    output += '\n';

    // Ears
    output += this.rainbows[1].join('');
    output += ' ';
    output += style.magenta(` |${tick ? ' .' : ' . '}`);
    output += style.gray('/\\_/\\');
    output += '\n';

    // Face
    output += this.rainbows[2].join('');
    output += ' ';

    if (tick) {
      output += style.gray('~');
      output += style.magenta('|_');
    } else {
      output += style.gray('^');
      output += style.magenta('|__');
    }

    output += style.gray(this.getCatFace());
    output += '\n';

    // Feet
    output += this.rainbows[3].join('');
    output += ' ';

    if (tick) {
      output += style.gray(' "  ""  "');
    } else {
      output += style.gray('  ""  ""');
    }

    output += '\n';

    if (this.isFinalRender()) {
      return output;
    }

    // Routine
    const routine = this.activeRoutine;
    const task = this.activeTask;
    let line = '\n';

    if (routine) {
      line += this.style(routine.key.toUpperCase(), this.getColorType(routine), ['bold']);
      line += ' ';
      line += routine.title;
    }

    if (task) {
      line += ' ';
      line += this.style(task.statusText || task.title, 'pending');
    }

    line += '\n';
    output += this.truncate(line);

    return output;
  }
}
