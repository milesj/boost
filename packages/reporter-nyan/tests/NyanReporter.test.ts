import { Tool, STATUS_FAILED, STATUS_PASSED, STATUS_PENDING } from '@boost/core';
import { mockTool, mockConsole, mockRoutine, mockTask } from '@boost/test-utils';
import NyanReporter from '../src/NyanReporter';

describe('NyanReporter', () => {
  let reporter: NyanReporter;
  let tool: Tool<any>;

  beforeEach(() => {
    jest.useFakeTimers();

    tool = mockTool();

    reporter = new NyanReporter();
    reporter.console = mockConsole(tool);
    reporter.tool = tool;
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  describe('bootstrap()', () => {
    it('binds events', () => {
      const spy = jest.spyOn(reporter.console, 'on');

      reporter.bootstrap();

      expect(spy).toHaveBeenCalledWith('start', expect.anything());
      expect(spy).toHaveBeenCalledWith('stop', expect.anything());
      expect(spy).toHaveBeenCalledWith('routine', expect.anything());
      expect(spy).toHaveBeenCalledWith('routine.pass', expect.anything());
      expect(spy).toHaveBeenCalledWith('routine.fail', expect.anything());
      expect(spy).toHaveBeenCalledWith('task', expect.anything());
    });

    it('generates rainbow data', () => {
      expect(reporter.rainbowWidth).toBe(0);
      expect(reporter.rainbowColors).toEqual([]);

      reporter.bootstrap();

      expect(reporter.rainbowWidth).not.toBe(0);
      expect(reporter.rainbowColors).not.toEqual([]);
    });
  });

  describe('handleStart()', () => {
    it('enqueues an output', () => {
      expect(reporter.console.outputQueue).toEqual([]);

      reporter.handleStart();

      expect(reporter.console.outputQueue).not.toEqual([]);
    });

    it('hides the cursor', () => {
      const spy = jest.spyOn(reporter.console, 'hideCursor');

      reporter.handleStart();

      expect(spy).toHaveBeenCalled();
    });
  });

  describe('handleStop()', () => {
    it('sets `failed` flag', () => {
      reporter.handleStop(null);

      expect(reporter.failed).toBe(false);

      reporter.handleStop(new Error());

      expect(reporter.failed).toBe(true);
    });

    it('shows the cursor', () => {
      const spy = jest.spyOn(reporter.console, 'showCursor');

      reporter.handleStop(null);

      expect(spy).toHaveBeenCalled();
    });
  });

  describe('handleRoutine()', () => {
    it('sets active routine', () => {
      const routine = mockRoutine(tool);

      expect(reporter.activeRoutine).toBeNull();

      reporter.handleRoutine(routine);

      expect(reporter.activeRoutine).toBe(routine);
    });
  });

  describe('handleTask()', () => {
    it('sets active task', () => {
      const task = mockTask();

      expect(reporter.activeTask).toBeNull();

      reporter.handleTask(task);

      expect(reporter.activeTask).toBe(task);
    });
  });

  describe('getCatFace()', () => {
    it('returns a normal face', () => {
      expect(reporter.getCatFace()).toMatchSnapshot();
    });

    it('returns a cross eyes face when failed', () => {
      reporter.failed = true;

      expect(reporter.getCatFace()).toMatchSnapshot();
    });

    it('returns a cross eyes face when routine is failed', () => {
      reporter.activeRoutine = mockRoutine(tool);
      reporter.activeRoutine!.status = STATUS_FAILED;

      expect(reporter.getCatFace()).toMatchSnapshot();
    });

    it('returns a happy eyes face when routine is passed', () => {
      reporter.activeRoutine = mockRoutine(tool);
      reporter.activeRoutine!.status = STATUS_PASSED;

      expect(reporter.getCatFace()).toMatchSnapshot();
    });

    it('returns an open eyes face when routine is pending', () => {
      reporter.activeRoutine = mockRoutine(tool);
      reporter.activeRoutine!.status = STATUS_PENDING;

      expect(reporter.getCatFace()).toMatchSnapshot();
    });

    it('returns an open eyes face when routine is skipped', () => {
      reporter.activeRoutine = mockRoutine(tool).skip();

      expect(reporter.getCatFace()).toMatchSnapshot();
    });
  });

  describe('increaseRainbowWidth()', () => {
    beforeEach(() => {
      reporter.bootstrap();
    });

    it('increases width each call', () => {
      reporter.increaseRainbowWidth();

      expect(reporter.rainbows).toMatchSnapshot();

      reporter.increaseRainbowWidth();

      expect(reporter.rainbows).toMatchSnapshot();

      reporter.increaseRainbowWidth();

      expect(reporter.rainbows).toMatchSnapshot();
    });

    it('doesnt go past max width', () => {
      reporter.rainbowWidth = 3;
      reporter.increaseRainbowWidth();
      reporter.increaseRainbowWidth();
      reporter.increaseRainbowWidth();
      reporter.increaseRainbowWidth();
      reporter.increaseRainbowWidth();

      expect(reporter.rainbows[0]).toHaveLength(3);
    });
  });

  describe('renderLines()', () => {
    beforeEach(() => {
      reporter.bootstrap();
    });

    it('renders base cat', () => {
      expect(reporter.renderLines()).toMatchSnapshot();
    });

    it('renders with alternative ticks', () => {
      expect(reporter.renderLines()).toMatchSnapshot();
      expect(reporter.renderLines()).toMatchSnapshot();
      expect(reporter.renderLines()).toMatchSnapshot();
      expect(reporter.renderLines()).toMatchSnapshot();
      expect(reporter.renderLines()).toMatchSnapshot();
    });

    it('renders with a routine', () => {
      reporter.activeRoutine = mockRoutine(tool);

      expect(reporter.renderLines()).toMatchSnapshot();
    });

    it('renders with a routine and task', () => {
      reporter.activeRoutine = mockRoutine(tool);
      reporter.activeTask = mockTask();

      expect(reporter.renderLines()).toMatchSnapshot();
    });

    it('doesnt render routine and task when final render', () => {
      reporter.activeRoutine = mockRoutine(tool);
      reporter.activeTask = mockTask();
      // @ts-ignore
      reporter.console.state.final = true;

      expect(reporter.renderLines()).toMatchSnapshot();
    });
  });
});
