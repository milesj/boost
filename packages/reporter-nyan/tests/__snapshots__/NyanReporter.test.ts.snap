// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NyanReporter getCatFace() returns a cross eyes face when failed 1`] = `"( [37mx .x[39m)"`;

exports[`NyanReporter getCatFace() returns a cross eyes face when routine is failed 1`] = `"( [37mx .x[39m)"`;

exports[`NyanReporter getCatFace() returns a happy eyes face when routine is passed 1`] = `"( [37m^ .^[39m)"`;

exports[`NyanReporter getCatFace() returns a normal face 1`] = `"( [37m- .-[39m)"`;

exports[`NyanReporter getCatFace() returns an open eyes face when routine is pending 1`] = `"( [37mo .o[39m)"`;

exports[`NyanReporter getCatFace() returns an open eyes face when routine is skipped 1`] = `"( [37mo .o[39m)"`;

exports[`NyanReporter increaseRainbowWidth() increases width each call 1`] = `
Array [
  Array [
    "[38;5;154m-[0m",
  ],
  Array [
    "[38;5;154m-[0m",
  ],
  Array [
    "[38;5;154m-[0m",
  ],
  Array [
    "[38;5;154m-[0m",
  ],
]
`;

exports[`NyanReporter increaseRainbowWidth() increases width each call 2`] = `
Array [
  Array [
    "[38;5;154m-[0m",
    "[38;5;154m-[0m",
  ],
  Array [
    "[38;5;154m-[0m",
    "[38;5;154m-[0m",
  ],
  Array [
    "[38;5;154m-[0m",
    "[38;5;154m-[0m",
  ],
  Array [
    "[38;5;154m-[0m",
    "[38;5;154m-[0m",
  ],
]
`;

exports[`NyanReporter increaseRainbowWidth() increases width each call 3`] = `
Array [
  Array [
    "[38;5;154m-[0m",
    "[38;5;154m-[0m",
    "[38;5;154m-[0m",
  ],
  Array [
    "[38;5;154m-[0m",
    "[38;5;154m-[0m",
    "[38;5;154m-[0m",
  ],
  Array [
    "[38;5;154m-[0m",
    "[38;5;154m-[0m",
    "[38;5;154m-[0m",
  ],
  Array [
    "[38;5;154m-[0m",
    "[38;5;154m-[0m",
    "[38;5;154m-[0m",
  ],
]
`;

exports[`NyanReporter renderLines() doesnt render routine and task when final render 1`] = `
"
[38;5;154m-[0m [35m ,------,[39m
[38;5;154m-[0m [35m | . [39m[90m/\\\\_/\\\\[39m
[38;5;154m-[0m [90m^[39m[35m|__[39m[90m( [37mo .o[90m)[39m
[38;5;154m-[0m [90m  \\"\\"  \\"\\"[39m
"
`;

exports[`NyanReporter renderLines() renders base cat 1`] = `
"
[38;5;154m-[0m [35m ,------,[39m
[38;5;154m-[0m [35m | . [39m[90m/\\\\_/\\\\[39m
[38;5;154m-[0m [90m^[39m[35m|__[39m[90m( [37m- .-[90m)[39m
[38;5;154m-[0m [90m  \\"\\"  \\"\\"[39m


"
`;

exports[`NyanReporter renderLines() renders with a routine 1`] = `
"
[38;5;154m-[0m [35m ,------,[39m
[38;5;154m-[0m [35m | . [39m[90m/\\\\_/\\\\[39m
[38;5;154m-[0m [90m^[39m[35m|__[39m[90m( [37mo .o[90m)[39m
[38;5;154m-[0m [90m  \\"\\"  \\"\\"[39m

[90m[1mKEY[22m[39m Title
"
`;

exports[`NyanReporter renderLines() renders with a routine and task 1`] = `
"
[38;5;154m-[0m [35m ,------,[39m
[38;5;154m-[0m [35m | . [39m[90m/\\\\_/\\\\[39m
[38;5;154m-[0m [90m^[39m[35m|__[39m[90m( [37mo .o[90m)[39m
[38;5;154m-[0m [90m  \\"\\"  \\"\\"[39m

[90m[1mKEY[22m[39m Title [90mTitle[39m
"
`;

exports[`NyanReporter renderLines() renders with alternative ticks 1`] = `
"
[38;5;154m-[0m [35m ,------,[39m
[38;5;154m-[0m [35m | . [39m[90m/\\\\_/\\\\[39m
[38;5;154m-[0m [90m^[39m[35m|__[39m[90m( [37m- .-[90m)[39m
[38;5;154m-[0m [90m  \\"\\"  \\"\\"[39m


"
`;

exports[`NyanReporter renderLines() renders with alternative ticks 2`] = `
"
[38;5;154m-[0m[38;5;154m_[0m [35m ,------,[39m
[38;5;154m-[0m[38;5;154m_[0m [35m | .[39m[90m/\\\\_/\\\\[39m
[38;5;154m-[0m[38;5;154m_[0m [90m~[39m[35m|_[39m[90m( [37m- .-[90m)[39m
[38;5;154m-[0m[38;5;154m_[0m [90m \\"  \\"\\"  \\"[39m


"
`;

exports[`NyanReporter renderLines() renders with alternative ticks 3`] = `
"
[38;5;154m-[0m[38;5;154m_[0m[38;5;154m_[0m [35m ,------,[39m
[38;5;154m-[0m[38;5;154m_[0m[38;5;154m_[0m [35m | .[39m[90m/\\\\_/\\\\[39m
[38;5;154m-[0m[38;5;154m_[0m[38;5;154m_[0m [90m~[39m[35m|_[39m[90m( [37m- .-[90m)[39m
[38;5;154m-[0m[38;5;154m_[0m[38;5;154m_[0m [90m \\"  \\"\\"  \\"[39m


"
`;

exports[`NyanReporter renderLines() renders with alternative ticks 4`] = `
"
[38;5;154m-[0m[38;5;154m_[0m[38;5;154m_[0m[38;5;184m-[0m [35m ,------,[39m
[38;5;154m-[0m[38;5;154m_[0m[38;5;154m_[0m[38;5;184m-[0m [35m | . [39m[90m/\\\\_/\\\\[39m
[38;5;154m-[0m[38;5;154m_[0m[38;5;154m_[0m[38;5;184m-[0m [90m^[39m[35m|__[39m[90m( [37m- .-[90m)[39m
[38;5;154m-[0m[38;5;154m_[0m[38;5;154m_[0m[38;5;184m-[0m [90m  \\"\\"  \\"\\"[39m


"
`;

exports[`NyanReporter renderLines() renders with alternative ticks 5`] = `
"
[38;5;154m-[0m[38;5;154m_[0m[38;5;154m_[0m[38;5;184m-[0m[38;5;184m-[0m [35m ,------,[39m
[38;5;154m-[0m[38;5;154m_[0m[38;5;154m_[0m[38;5;184m-[0m[38;5;184m-[0m [35m | . [39m[90m/\\\\_/\\\\[39m
[38;5;154m-[0m[38;5;154m_[0m[38;5;154m_[0m[38;5;184m-[0m[38;5;184m-[0m [90m^[39m[35m|__[39m[90m( [37m- .-[90m)[39m
[38;5;154m-[0m[38;5;154m_[0m[38;5;154m_[0m[38;5;184m-[0m[38;5;184m-[0m [90m  \\"\\"  \\"\\"[39m


"
`;
